File: access/deploy/cloudflare-worker/src/temporal-comparison.ts
/** Compare exact source date envelopes, without parsing or historical judgment. */
import { HttpError, stringArray, type Item } from './common.ts';
import { KnowledgeRevisionConflict } from './lens-pagination.ts';
import {NativeBudgetExceeded, nativeNumberInfo, pythonEquals, pythonMember, codePointCompare} from '../../../shared/native-semantics.ts';
import {nativeStrip} from '../../../shared/native-unicode.ts';
import {nativeChild, nativeField, nativeKeys, parseNativeJson, nativePacketObject, derived,
  type NativeRef, type NativePacket} from './native-lens.ts';
import {canonicalNativeJson} from './native-source-target.ts';

type Selection = { node_id: string; content_revision: string };
export type TemporalRequest = {
  schema_version: 'tos_temporal_comparison_request_v1'; source_revision: string;
  left: Selection; right: Selection;
};
type Status = 'comparable' | 'undetermined' | 'unsupported';
type Operand = { claim: NativeRef; value: NativeRef | null; normalized_time: NativeRef | null };
type CheckedOperand = { packet: Operand; status: Status; issues: string[] };
type Lookup = (identifier: string) => Promise<NativeRef[]>;
type Reason = { side: 'left' | 'right' | 'pair'; code: string };
const MAX_SOURCE_BYTES = 262144;

export function temporalNodeFromJson(text: string): NativeRef {
  try {
    const ref = parseNativeJson(text);
    if (!object(ref.value)) throw new Error('selected normalized carrier is not an object');
    return ref;
  } catch {throw new HttpError(503, 'selected normalized carrier is invalid');}
}

function object(value: unknown): Item | null {
  return value !== null && typeof value === 'object' && !Array.isArray(value) ? value as Item : null;
}
function fields(value: unknown): Item { return object(value) ?? {}; }
function requireOperandContainers(node: Item): void {
  if (['attributes', 'semantics', 'type_mapping'].some(key => !object(node[key]))) {
    throw new HttpError(503, 'selected normalized carrier has invalid structural containers');
  }
  const semantics = fields(node.semantics);
  if (['claim', 'time'].some(key => Object.hasOwn(semantics, key) && !object(semantics[key]))) {
    throw new HttpError(503, 'selected normalized carrier has invalid structural containers');
  }
}
function exactKeys(value: Item, keys: string[]): boolean {
  return Object.keys(value).length === keys.length && keys.every(key => Object.hasOwn(value, key));
}
function revision(value: unknown): value is string {
  return typeof value === 'string' && /^[a-f0-9]{64}$/.test(value);
}

export function normalizeTemporalComparisonRequest(value: unknown): TemporalRequest {
  const request = object(value);
  if (!request || !exactKeys(request, ['schema_version', 'source_revision', 'left', 'right'])) {
    throw new HttpError(400, 'temporal comparison requires schema_version, source_revision, left and right only');
  }
  if (request.schema_version !== 'tos_temporal_comparison_request_v1') throw new HttpError(400, 'unsupported temporal comparison request schema');
  if (!revision(request.source_revision)) throw new HttpError(400, 'source_revision must be an exact knowledge snapshot revision');
  function selection(side: 'left' | 'right'): Selection {
    const ref = object(request![side]);
    if (!ref || !exactKeys(ref, ['node_id', 'content_revision'])) throw new HttpError(400, `${side} requires node_id and content_revision only`);
    const id = ref.node_id;
    // Count Unicode code points, as the public schema/Python reader do.
    if (typeof id !== 'string' || [...id].length < 1 || [...id].length > 1024 || nativeStrip(id) !== id) {
      throw new HttpError(400, `${side}.node_id must be a nonempty exact normalized ID of at most 1024 characters`);
    }
    if (!revision(ref.content_revision)) throw new HttpError(400, `${side}.content_revision must be an exact carrier revision`);
    // Python dict(ref) retains the accepted selection's member order.
    return Object.keys(ref)[0] === 'node_id'
      ? {node_id:id,content_revision:ref.content_revision}
      : {content_revision:ref.content_revision,node_id:id};
  }
  return { schema_version: 'tos_temporal_comparison_request_v1', source_revision: request.source_revision,
    left: selection('left'), right: selection('right') };
}

/** Temporal JSON equality deliberately excludes Python's bool/int alias. */
function sameJson(left: NativeRef, right: NativeRef): boolean {
  let visits = 300000;
  function equal(a: NativeRef, b: NativeRef): boolean {
    if (--visits < 0) throw new NativeBudgetExceeded('temporal equality visit budget');
    if (typeof a.value === 'boolean' || typeof b.value === 'boolean') return a.value === b.value;
    if (typeof a.value === 'number' && typeof b.value === 'number') return pythonEquals(a,b);
    if (typeof a.value !== typeof b.value || (a.value === null) !== (b.value === null)) return false;
    if (!a.value || typeof a.value !== 'object') return a.value === b.value;
    if (Array.isArray(a.value) !== Array.isArray(b.value)) return false;
    const keys = nativeKeys(a), other = new Set(nativeKeys(b));
    return keys.length === other.size && keys.every(key => other.has(key) && equal(nativeChild(a,key),nativeChild(b,key)));
  }
  return equal(left,right);
}

function safeInteger(ref: NativeRef): boolean {
  return typeof ref.value === 'number' && Number.isSafeInteger(ref.value);
}

async function hashText(text: string): Promise<string> {
  const digest = await crypto.subtle.digest('SHA-256', new TextEncoder().encode(text));
  return Array.from(new Uint8Array(digest), byte => byte.toString(16).padStart(2, '0')).join('');
}

async function documentCatalogueBinding(claimRef: NativeRef, valueRef: NativeRef, lookup: Lookup): Promise<string | null> {
  const claim = fields(claimRef.value), value = fields(valueRef.value);
  const sourceRef = nativeField(claimRef,'attributes.source_claim'), source = fields(sourceRef.value);
  const semanticsRef = nativeField(claimRef,'semantics.claim'), semantics = fields(semanticsRef.value);
  const timeRef = nativeField(valueRef,'semantics.time'), time = fields(timeRef.value);
  const profile = fields(semantics.source_claim_profile), schemas = profile.schemas;
  const attributionRef = nativeField(sourceRef,'qualifiers.catalogue_attribution');
  const attribution = fields(attributionRef.value), rawRef = nativeField(sourceRef,'object'), raw = fields(rawRef.value);
  if (source.predicate !== 'document_catalogue_date' || source.schema_version !== 'tos_document_catalogue_claim_v1'
      || source.assertion_layer !== 'bibliographic_assertion' || semantics.relation_type_id !== 'tos.relation.document-catalogue-date'
      || profile.reader !== 'document-catalogue-temporal-v1'
      || !sameJson(nativeField(semanticsRef,'source_claim_profile.assertion_layers'),parseNativeJson('["bibliographic_assertion"]'))
      || !Array.isArray(schemas) || schemas.length !== 1 || !object(schemas[0])
      || schemas[0].schema_version !== source.schema_version || schemas[0].schema_ref !== 'ToS/contracts/document-catalogue-claim.schema.json'
      || raw.role !== 'catalogue-assigned-document-date' || time.role !== raw.role
      || typeof raw.kind !== 'string' || !['date-assertion', 'interval-assertion', 'unknown-date'].includes(raw.kind)
      || attribution.field_role !== 'assigned-date' || typeof attribution.source_field !== 'string' || !nativeStrip(attribution.source_field)
      || !Array.isArray(source.evidence_refs) || !pythonMember(nativeField(attributionRef,'evidence_ref'),nativeField(sourceRef,'evidence_refs'))
      || !sameJson(nativeField(attributionRef,'source_wording'),nativeField(rawRef,'source_wording'))) return 'document-catalogue-profile-binding-inconsistent';
  const subjects = typeof semantics.subject_node_id === 'string' ? await lookup(semantics.subject_node_id) : [];
  const subject = subjects.length === 1 ? fields(subjects[0]!.value) : {};
  if (subjects.length !== 1 || subject.entity_id !== source.subject_ref || subject.source_graph !== 'source-claims'
      || fields(subject.type_mapping).status !== 'mapped'
      || !stringArray(fields(subject.semantics).type_ancestors).includes('tos.entity.document')) {
    return 'document-catalogue-subject-binding-inconsistent';
  }
  const inconsistent = 'document-catalogue-exact-source-binding-inconsistent';
  const sourceText = canonicalNativeJson(sourceRef), rawText = canonicalNativeJson(rawRef);
  if (new TextEncoder().encode(sourceText).length > MAX_SOURCE_BYTES || semantics.source_canonical_json !== sourceText) return inconsistent;
  const digest = await hashText(sourceText), valueDigest = await hashText(rawText);
  const literalDigest = await hashText('{"claim_ref":' + canonicalNativeJson(nativeField(sourceRef,'claim_id')) + ',"value":' + rawText + '}');
  const left = fields(claim.attributes), right = fields(value.attributes);
  const sourceLine = nativeField(claimRef,'attributes.source_line');
  if (left.source_sha256 !== digest || right.source_sha256 !== digest || right.value_sha256 !== valueDigest
      || await hashText(canonicalNativeJson(nativeField(valueRef,'attributes.value'))) !== valueDigest
      || await hashText(canonicalNativeJson(nativeField(timeRef,'raw'))) !== valueDigest
      || value.native_id !== 'literal:sha256:' + literalDigest
      || typeof sourceLine.value !== 'number' || nativeNumberInfo(sourceLine).kind !== 'int'
      || BigInt(nativeNumberInfo(sourceLine).lexeme) < 1n
      || !pythonEquals(sourceLine,nativeField(valueRef,'attributes.source_line'))
      || !pythonEquals(nativeField(claimRef,'source_refs'),nativeField(valueRef,'source_refs'))) return inconsistent;
  return null;
}

async function operand(ref: Selection, lookup: Lookup): Promise<CheckedOperand> {
  const matches = await lookup(ref.node_id);
  if (matches.length !== 1) throw new HttpError(404, `expected one exact knowledge node: ${ref.node_id}`);
  const claimRef = matches[0]!, claim = fields(claimRef.value);
  if (claim.content_revision !== ref.content_revision) throw new KnowledgeRevisionConflict('selected Claim content changed; select again from the current snapshot');
  requireOperandContainers(claim);
  const packet: Operand = { claim: claimRef, value: null, normalized_time: null };
  const stop = (status: Status, issues: string[]): CheckedOperand => ({ packet, status, issues });
  if (claim.source_graph !== 'source-claims' || claim.kind_id !== 'claim' || claim.type_id !== 'tos.entity.claim'
      || fields(claim.type_mapping).status !== 'mapped') return stop('unsupported', ['selected-node-is-not-a-source-claim']);
  const semantics = fields(fields(claim.semantics).claim);
  const sourceRef = nativeField(claimRef,'attributes.source_claim');
  const source = object(sourceRef.value);
  if (!source || typeof source.claim_id !== 'string' || source.claim_id !== semantics.claim_id
      || !safeInteger(nativeField(sourceRef,'claim_version')) || Number(source.claim_version) < 1
      || !safeInteger(nativeField(claimRef,'semantics.claim.claim_version')) || source.claim_version !== semantics.claim_version
      || !pythonEquals(nativeField(sourceRef,'predicate'),nativeField(claimRef,'semantics.claim.source_predicate_id')) || semantics.predicate_mapping_status !== 'mapped') {
    return stop('undetermined', ['claim-source-binding-inconsistent']);
  }
  if (typeof semantics.object_node_id !== 'string') return stop('undetermined', ['claim-object-binding-unavailable']);
  const values = await lookup(semantics.object_node_id);
  if (values.length !== 1) return stop('undetermined', ['claim-object-unavailable-or-ambiguous']);
  const valueRef = values[0]!, value = fields(valueRef.value);
  requireOperandContainers(value);
  packet.value = valueRef;
  if (value.source_graph !== 'source-claims' || value.type_id !== 'tos.entity.temporal-assertion'
      || fields(value.type_mapping).status !== 'mapped') return stop('unsupported', ['claim-object-is-not-a-declared-temporal-assertion']);
  const attributes = fields(value.attributes);
  const documentary = source.predicate === 'document_catalogue_date' || source.schema_version === 'tos_document_catalogue_claim_v1'
    || fields(fields(value.semantics).time).role === 'catalogue-assigned-document-date';
  if (attributes.claim_ref !== source.claim_id || !Object.hasOwn(attributes, 'value') || (!documentary && !sameJson(nativeField(valueRef,'attributes.value'),nativeField(sourceRef,'object')))) {
    return stop('undetermined', ['temporal-object-source-binding-inconsistent']);
  }
  const time = object(fields(value.semantics).time);
  if (!time) return stop('undetermined', ['temporal-normalization-unavailable']);
  packet.normalized_time = nativeField(valueRef,'semantics.time');
  if (!Object.hasOwn(time, 'raw') || (!documentary && !sameJson(nativeField(valueRef,'semantics.time.raw'),nativeField(valueRef,'attributes.value')))) return stop('undetermined', ['temporal-normalization-source-binding-inconsistent']);
  if (documentary) {
    const issue = await documentCatalogueBinding(claimRef, valueRef, lookup);
    if (issue) return stop('undetermined', [issue]);
  }
  const rawIssues = time.issues ?? [];
  if ((Object.hasOwn(time, 'issues') && time.issues === null) || !Array.isArray(rawIssues)
      || rawIssues.some(issue => typeof issue !== 'string' || !issue)) return stop('unsupported', ['temporal-normalization-issues-invalid']);
  const issues = stringArray(rawIssues);
  if (typeof time.kind !== 'string' || !['date-assertion', 'interval-assertion', 'relative-order', 'unknown-date'].includes(time.kind)) {
    return stop('unsupported', ['unsupported-temporal-normalization-kind']);
  }
  const unsupported = issues.filter(issue => issue.startsWith('conflicting-') || ['reversed-interval', 'invalid-date-parts', 'unparsed-date-value'].includes(issue));
  if (time.calendar !== undefined && time.calendar !== null && (typeof time.calendar !== 'string'
      || !['gregorian', 'proleptic-gregorian'].includes(time.calendar))) unsupported.push('unsupported-declared-calendar');
  if (time.declared_year_numbering !== undefined && time.declared_year_numbering !== null && time.declared_year_numbering !== 'astronomical') unsupported.push('unsupported-declared-year-numbering');
  if (unsupported.length) return stop('unsupported', [...new Set([...issues, ...unsupported])]);
  if (!['date-assertion', 'interval-assertion'].includes(time.kind)) issues.push('no-absolute-date-bounds');
  if (time.calendar === undefined || time.calendar === null) issues.push('declared-calendar-unavailable');
  if (time.declared_year_numbering === undefined || time.declared_year_numbering === null) issues.push('declared-year-numbering-unavailable');
  if (typeof time.precision === 'string' && ['approximate', 'uncertain', 'unknown'].includes(time.precision)) issues.push('non-exact-date-precision');
  if (time.certainty !== 'exact') issues.push('explicit-exact-certainty-unavailable');
  if (time.comparison_calendar !== 'proleptic-gregorian') issues.push('comparison-calendar-unavailable');
  if (time.year_numbering !== 'astronomical') issues.push('comparison-year-numbering-unavailable');
  if (!Number.isSafeInteger(time.sort_start) || !Number.isSafeInteger(time.sort_end)) issues.push('absolute-date-envelope-unavailable');
  else if (Number(time.sort_start) > Number(time.sort_end)) issues.push('reversed-date-envelope');
  return stop(issues.length ? 'undetermined' : 'comparable', [...new Set(issues)]);
}

function envelopeRelation(left: Item, right: Item): string {
  const a = Number(left.sort_start), b = Number(left.sort_end), c = Number(right.sort_start), d = Number(right.sort_end);
  if (b < c) return 'before';
  if (d < a) return 'after';
  if (a === c && b === d) return 'equal';
  if (a <= c && b >= d) return 'contains';
  if (c <= a && d >= b) return 'contained-by';
  return 'overlaps';
}

export async function compareTemporalOperands(sourceRevision: unknown, requestValue: unknown, lookup: Lookup): Promise<NativePacket> {
  const request = normalizeTemporalComparisonRequest(requestValue);
  if (request.source_revision !== sourceRevision) throw new KnowledgeRevisionConflict('knowledge snapshot changed; select both Claims again');
  const left = await operand(request.left, lookup), right = await operand(request.right, lookup);
  const statuses = [left.status, right.status];
  let status: Status = statuses.includes('unsupported') ? 'unsupported' : statuses.includes('undetermined') ? 'undetermined' : 'comparable';
  const reasons: Reason[] = [...left.issues.map(code => ({ side: 'left' as const, code })), ...right.issues.map(code => ({ side: 'right' as const, code }))];
  let relation: string | null = null;
  if (status === 'comparable') {
    const leftTime = fields(left.packet.normalized_time!.value), rightTime = fields(right.packet.normalized_time!.value);
    if (typeof leftTime.role !== 'string' || !leftTime.role || typeof rightTime.role !== 'string' || !rightTime.role) {
      status = 'undetermined'; reasons.push({ side: 'pair', code: 'time-role-unavailable' });
    } else if (leftTime.role !== rightTime.role) {
      status = 'unsupported'; reasons.push({ side: 'pair', code: 'different-time-roles' });
    } else if (!['historical-time', 'catalogue-assigned-document-date'].includes(leftTime.role)) {
      status = 'unsupported'; reasons.push({ side: 'pair', code: 'unsupported-time-role' });
    } else relation = envelopeRelation(leftTime, rightTime);
  }
  const refs = [...new Set([left.packet.claim, left.packet.value, right.packet.claim, right.packet.value]
    .flatMap(item => {
      const values = item ? fields(item.value).source_refs : undefined;
      return Array.isArray(values) ? values.filter((value): value is string => typeof value === 'string') : [];
    }))].sort(codePointCompare);
  // Python's final ensure_ascii=False UTF-8 response rejects escaped lone
  // surrogates too. Validate only returned source carriers, not an unreturned
  // documentary subject's unrelated fields. No source values are rewritten.
  const seen = new WeakSet<object>();
  function validUtf8(ref: NativeRef): void {
    if (typeof ref.value === 'string' && !ref.value.isWellFormed()) throw new HttpError(503,'prepared response contains invalid JSON values');
    if (!ref.value || typeof ref.value !== 'object' || seen.has(ref.value)) return;
    seen.add(ref.value);
    for (const key of nativeKeys(ref)) {
      if (!key.isWellFormed()) throw new HttpError(503,'prepared response contains invalid JSON values');
      validUtf8(nativeChild(ref,key));
    }
  }
  for (const value of [left.packet.claim,left.packet.value,right.packet.claim,right.packet.value]) if (value) validUtf8(value);
  const packet = (operand: Operand) => nativePacketObject([
    ['claim',operand.claim],['value',operand.value],['normalized_time',operand.normalized_time],
  ]);
  return nativePacketObject([
    ['schema_version','tos_temporal_comparison_result_v1'], ['source_revision',request.source_revision], ['request',derived(request)],
    ['comparison',derived({status,relation,reasons,basis:'normalized-source-date-envelopes'})],
    ['left',packet(left.packet)], ['right',packet(right.packet)], ['source_refs',derived(refs)],
    ['authority_boundary',derived({is_source:false,writes_to_tree:false,performs_assessment:false,
      creates_inferred_claim:false,comparison_basis:'normalized-source-date-envelopes',
      note:'Relations describe date envelopes only. They do not establish event simultaneity, duration, causality, identity or the truth/admission of either Claim. Numeric keys are ordering keys, not timestamps or elapsed-time quantities.'})],
  ]);
}

File: access/src/tos_access/temporal_comparison.py
"""Compare two exact, source-declared date envelopes, never historical truth.

The owner normalizer supplies comparison keys. This reader does not parse
prose, convert calendars, follow relative anchors or estimate uncertainty.
Only exact index lookups are used after the shared graph index is prepared.
"""
from __future__ import annotations

import copy
import math
import re

from .lens_pagination import KnowledgeRevisionConflict


REQUEST_SCHEMA = 'tos_temporal_comparison_request_v1'
RESULT_SCHEMA = 'tos_temporal_comparison_result_v1'
SUPPORTED_TIME_ROLES = frozenset({'historical-time'})


class TemporalReadModelInvalid(RuntimeError):
    """The selected projection cannot supply a structurally valid operand."""


BOUNDARY = {
    'is_source': False, 'writes_to_tree': False, 'performs_assessment': False,
    'creates_inferred_claim': False,
    'comparison_basis': 'normalized-source-date-envelopes',
    'note': 'Relations describe date envelopes only. They do not establish event '
            'simultaneity, duration, causality, identity or the truth/admission of either Claim. '
            'Numeric keys are ordering keys, not timestamps or elapsed-time quantities.',
}


def normalize_temporal_comparison_request(value):
    if not isinstance(value, dict) or set(value) != {'schema_version', 'source_revision', 'left', 'right'}:
        raise ValueError('temporal comparison requires schema_version, source_revision, left and right only')
    if value['schema_version'] != REQUEST_SCHEMA:
        raise ValueError('unsupported temporal comparison request schema')
    result = {'schema_version': REQUEST_SCHEMA}
    revision = value['source_revision']
    if not isinstance(revision, str) or not re.fullmatch('[0-9a-f]{64}', revision):
        raise ValueError('source_revision must be an exact knowledge snapshot revision')
    result['source_revision'] = revision
    for side in ('left', 'right'):
        ref = value[side]
        if not isinstance(ref, dict) or set(ref) != {'node_id', 'content_revision'}:
            raise ValueError(f'{side} requires node_id and content_revision only')
        identifier = ref['node_id']
        if not isinstance(identifier, str) or not 1 <= len(identifier) <= 1024 or identifier != identifier.strip():
            raise ValueError(f'{side}.node_id must be a nonempty exact normalized ID of at most 1024 characters')
        if not isinstance(ref['content_revision'], str) or not re.fullmatch('[0-9a-f]{64}', ref['content_revision']):
            raise ValueError(f'{side}.content_revision must be an exact carrier revision')
        result[side] = dict(ref)
    return result


def _same_json(left, right):
    # Compare JSON values, not serialization spelling: 1 and 1.0 are numbers,
    # but True and 1 must never alias through Python equality. Exact source
    # record digests remain in the returned carriers, separately from this join.
    if isinstance(left, bool) or isinstance(right, bool):
        return type(left) is type(right) and left == right
    if isinstance(left, (int, float)) and isinstance(right, (int, float)):
        return left == right
    if type(left) is not type(right):
        return False
    if isinstance(left, dict):
        return left.keys() == right.keys() and all(_same_json(left[key], right[key]) for key in left)
    if isinstance(left, list):
        return len(left) == len(right) and all(_same_json(a, b) for a, b in zip(left, right))
    return left == right


def _safe_integer(value):
    return type(value) in (int, float) and abs(value) <= 9007199254740991 and math.isfinite(value) and value == int(value)


def _fields(value):
    return value if isinstance(value, dict) else {}


def _require_operand_containers(node):
    # These known containers must obey the existing normalized-node ABI.
    # Reject damaged carriers, never scrub them into a successful packet.
    if any(not isinstance(node.get(key), dict) for key in ('attributes', 'semantics', 'type_mapping')):
        raise TemporalReadModelInvalid('selected normalized carrier has invalid structural containers')
    if any(key in node['semantics'] and not isinstance(node['semantics'][key], dict) for key in ('claim', 'time')):
        raise TemporalReadModelInvalid('selected normalized carrier has invalid structural containers')


def _operand(ref, lookup):
    matches = lookup(ref['node_id'])
    if len(matches) != 1:
        raise KeyError(f'expected one exact knowledge node: {ref["node_id"]}')
    claim = matches[0]
    if claim.get('content_revision') != ref['content_revision']:
        raise KnowledgeRevisionConflict('selected Claim content changed; select again from the current snapshot')
    _require_operand_containers(claim)
    packet = {'claim': copy.deepcopy(claim), 'value': None, 'normalized_time': None}
    if (claim.get('source_graph') != 'source-claims' or claim.get('kind_id') != 'claim'
            or claim.get('type_id') != 'tos.entity.claim' or _fields(claim.get('type_mapping')).get('status') != 'mapped'):
        return packet, 'unsupported', ['selected-node-is-not-a-source-claim']
    semantics = _fields(_fields(claim.get('semantics')).get('claim'))
    source = _fields(claim.get('attributes')).get('source_claim')
    if (not isinstance(source, dict) or not isinstance(source.get('claim_id'), str)
            or source['claim_id'] != semantics.get('claim_id')
            or not _safe_integer(source.get('claim_version')) or source['claim_version'] < 1
            or not _safe_integer(semantics.get('claim_version'))
            or not _same_json(source.get('claim_version'), semantics.get('claim_version'))
            or source.get('predicate') != semantics.get('source_predicate_id')
            or semantics.get('predicate_mapping_status') != 'mapped'):
        return packet, 'undetermined', ['claim-source-binding-inconsistent']
    identifier = semantics.get('object_node_id')
    if not isinstance(identifier, str):
        return packet, 'undetermined', ['claim-object-binding-unavailable']
    values = lookup(identifier)
    if len(values) != 1:
        return packet, 'undetermined', ['claim-object-unavailable-or-ambiguous']
    value = values[0]
    _require_operand_containers(value)
    packet['value'] = copy.deepcopy(value)
    if (value.get('source_graph') != 'source-claims' or value.get('type_id') != 'tos.entity.temporal-assertion'
            or _fields(value.get('type_mapping')).get('status') != 'mapped'):
        return packet, 'unsupported', ['claim-object-is-not-a-declared-temporal-assertion']
    attributes = _fields(value.get('attributes'))
    if attributes.get('claim_ref') != source['claim_id'] or 'value' not in attributes or not _same_json(attributes['value'], source.get('object')):
        return packet, 'undetermined', ['temporal-object-source-binding-inconsistent']
    time = _fields(value.get('semantics')).get('time')
    if not isinstance(time, dict):
        return packet, 'undetermined', ['temporal-normalization-unavailable']
    packet['normalized_time'] = copy.deepcopy(time)
    if 'raw' not in time or not _same_json(time['raw'], attributes['value']):
        return packet, 'undetermined', ['temporal-normalization-source-binding-inconsistent']
    raw_issues = time.get('issues', [])
    if not isinstance(raw_issues, list) or any(not isinstance(issue, str) or not issue for issue in raw_issues):
        return packet, 'unsupported', ['temporal-normalization-issues-invalid']
    issues = list(raw_issues)
    if time.get('kind') not in ('date-assertion', 'interval-assertion', 'relative-order', 'unknown-date'):
        return packet, 'unsupported', ['unsupported-temporal-normalization-kind']
    # An absent declaration is unknown; a declared unsupported system or an
    # invalid/conflicting shape is not an alternate coordinate conversion.
    unsupported = [issue for issue in issues if issue.startswith('conflicting-')
                   or issue in ('reversed-interval', 'invalid-date-parts', 'unparsed-date-value')]
    if time.get('calendar') not in (None, 'gregorian', 'proleptic-gregorian'):
        unsupported.append('unsupported-declared-calendar')
    if time.get('declared_year_numbering') not in (None, 'astronomical'):
        unsupported.append('unsupported-declared-year-numbering')
    if unsupported:
        return packet, 'unsupported', list(dict.fromkeys([*issues, *unsupported]))
    # Stale or malformed normalized carriers cannot substitute ready keys for
    # an explicit absolute-date kind and declared coordinate system.
    if time.get('kind') not in ('date-assertion', 'interval-assertion'):
        issues.append('no-absolute-date-bounds')
    if time.get('calendar') is None:
        issues.append('declared-calendar-unavailable')
    if time.get('declared_year_numbering') is None:
        issues.append('declared-year-numbering-unavailable')
    if time.get('precision') in ('approximate', 'uncertain', 'unknown'):
        issues.append('non-exact-date-precision')
    if time.get('certainty') != 'exact':
        issues.append('explicit-exact-certainty-unavailable')
    if time.get('comparison_calendar') != 'proleptic-gregorian':
        issues.append('comparison-calendar-unavailable')
    if time.get('year_numbering') != 'astronomical':
        issues.append('comparison-year-numbering-unavailable')
    if any(not _safe_integer(time.get(key)) for key in ('sort_start', 'sort_end')):
        issues.append('absolute-date-envelope-unavailable')
    elif time['sort_start'] > time['sort_end']:
        issues.append('reversed-date-envelope')
    if issues:
        return packet, 'undetermined', list(dict.fromkeys(issues))
    return packet, 'comparable', []


def _relation(left, right):
    a, b = left['sort_start'], left['sort_end']
    c, d = right['sort_start'], right['sort_end']
    if b < c:
        return 'before'
    if d < a:
        return 'after'
    if a == c and b == d:
        return 'equal'
    if a <= c and b >= d:
        return 'contains'
    if c <= a and d >= b:
        return 'contained-by'
    return 'overlaps'


def compare_temporal_operands(source_revision, request_value, lookup):
    """Transport-neutral computation over an immutable, exact-ID lookup.

Adapters bind the surrounding read snapshot. A lookup returns a sequence so
that missing and ambiguous IDs cannot be silently resolved by source priority.
    """
    request = normalize_temporal_comparison_request(request_value)
    if request['source_revision'] != source_revision:
        raise KnowledgeRevisionConflict('knowledge snapshot changed; select both Claims again')
    left, left_status, left_issues = _operand(request['left'], lookup)
    right, right_status, right_issues = _operand(request['right'], lookup)
    statuses = (left_status, right_status)
    status = 'unsupported' if 'unsupported' in statuses else 'undetermined' if 'undetermined' in statuses else 'comparable'
    reasons = [{'side': side, 'code': issue} for side, issues in (('left', left_issues), ('right', right_issues)) for issue in issues]
    relation = None
    if status == 'comparable':
        left_role, right_role = left['normalized_time'].get('role'), right['normalized_time'].get('role')
        if not isinstance(left_role, str) or not left_role or not isinstance(right_role, str) or not right_role:
            status = 'undetermined'
            reasons.append({'side': 'pair', 'code': 'time-role-unavailable'})
        elif left_role != right_role:
            status = 'unsupported'
            reasons.append({'side': 'pair', 'code': 'different-time-roles'})
        elif left_role not in SUPPORTED_TIME_ROLES:
            status = 'unsupported'
            reasons.append({'side': 'pair', 'code': 'unsupported-time-role'})
        else:
            relation = _relation(left['normalized_time'], right['normalized_time'])
    refs = sorted({ref for operand in (left, right) for item in (operand['claim'], operand['value'])
                   if item for ref in item.get('source_refs', []) if isinstance(ref, str)})
    return {'schema_version': RESULT_SCHEMA, 'source_revision': source_revision, 'request': request,
            'comparison': {'status': status, 'relation': relation, 'reasons': reasons,
                           'basis': 'normalized-source-date-envelopes'},
            'left': left, 'right': right, 'source_refs': refs, 'authority_boundary': dict(BOUNDARY)}


def compare_temporal_claims(graph, request, *, graph_index):
    graph_index.require_snapshot(graph)
    return compare_temporal_operands(graph['source_revision'], request,
                                     lambda identifier: graph_index.node_ids.get(identifier, ()))
